### ISSUE TITLE
Connection failure problems

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#1465]
Fix the bug that caused the lobby to fail with a useless error message (see screenshots).

<img width="1368" alt="Screenshot 2024-07-12 at 11 36 52" src="https://github.com/user-attachments/assets/8202a961-632e-47de-8ea0-bf39c028fd04">
<img width="1356" alt="Screenshot 2024-07-12 at 11 37 16" src="https://github.com/user-attachments/assets/800d2a41-d9a1-4aef-9f0b-8845ebe0b30c">

**Comment by ctm** [comment]
This appears to be pretty bad, and I think it's a regression. AFAICT, if we try to open a WebSocket, and it fails, the failure occurs at a level that prevents the failure from being reported back to rust through the normal channels.

We make a call to `wasm_sockets::EventClient::new`, which should return a `Result`. In the happy path, we get an `Ok` back and everything works. However, if something goes wrong, instead of getting a `Err(WebSocketError)` back, the program dies.  I do not think this is a problem with `wasm_sockets`. It appears to be coming from `web_sys::WebSocket`, which `wasm_sockets` uses, and that's just a `wasm-bindgen` wrapper.

I should be able to write a minimal test case that shows that `web_sys::WebSocket::new` can no longer return an `Err`, but I'm not sure that I'll be able to figure out what changed.

**Comment by ctm** [comment]
FWIW, `web_sys::WebSocket::new` is marked with `catch`:
```
    #[wasm_bindgen(catch, constructor, js_class = "WebSocket")]
    #[doc = "The `new WebSocket(..)` constructor, creating a new instance of `WebSocket`."]
    #[doc = ""]
    #[doc = "[MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket)"]
    #[doc = ""]
    #[doc = "*This API requires the following crate features to be activated: `WebSocket`*"]
    pub fn new(url: &str) -> Result<WebSocket, JsValue>;
```
IIRC, that means `wasm-bindgen` will automatically convert a failed open into an `Err`. This may simply be a regression in `wasm-bindgen`. OTOH, I don't see any open issues in the `wasm-bindgen` repository, so it probably isn't that simple. 

A minimal test case will be useful, but my development stack is certainly not minimal, so to create one, I'll need to research what the best minimal use of wasm-bindgen is. Trunk?

**Comment by ctm** [comment]
Looks like [`wasm-bindgen-test`](https://rustwasm.github.io/wasm-bindgen/wasm-bindgen-test/usage.html) may be sufficient.

**Comment by ctm** [comment]
There appear to be at least two separate issues. The biggest thing that took me by surprise (although I think I learned this long ago and then forgot it) is that the initial "WebSocket connection to ... failed" error on the console does not mean the app is dead from that point on. In fact, our on error handler is called with an `ErrorEvent`. Eventually `ReconnectingWebSocket::update` calls that `ErrorEvent`'s `message()` method to construct a string to tell the user what is going on.

That `message()` call fails, when I artificially cause the connection to fail by specifying a bad url. The failure I see is a complaint about `message()` throwing an error without bindgen labeling that method as `catch`, and the error itself is some complaint about expecting a string argument but getting undefined. Since `message()` doesn't take an argument and can't throw an error, I strongly suspect that we're not getting an `ErrorEvent` after all, probably due to memory shenanigans associated with the JavaScript Rust interface.

I can inspect some of the code associated with the chain of custody for that message, but this is not something I'm particularly well versed in. So, there's a chance I'll see a gotcha, but most likely I'll just spin my wheels.

Investigating further is apt to be slow and unrewarding, but here I go&hellip;

**Comment by ctm** [comment]
Specifically, the error we get from calling `message()` is:
```
index_bg.js:555 wasm-bindgen: imported JS function that was not marked as `catch` threw an error: expected a string argument, found undefined
```

**Comment by ctm** [comment]
So, calling the various `Event` methods (the first three) works fine:
```
                log::info!("type {}", error.type_());
                log::info!("time_stamp {}", error.time_stamp());
                log::info!("is_trusted {}", error.is_trusted());
                log::info!("filename: {}", error.filename());
```
```
index_bg.js:1450 INFO front-ends/web/src/reconnecting_web_socket.rs:148 type error
15:42:09.444 index_bg.js:1450 INFO front-ends/web/src/reconnecting_web_socket.rs:149 time_stamp 9700
15:42:09.445 index_bg.js:1450 INFO front-ends/web/src/reconnecting_web_socket.rs:150 is_trusted true
[error from calling filename elided, but it's basically the same as from calling message()]
```

The `filename()` method is specific to `ErrorEvent`, whereas `type_()`, `time_stamp()` and `is_trusted()` are all specific to `Event`, but the `type_()` response is "error".

So, AFAICT, we really do have an `ErrorEvent`. I just don't understand why we can't call `message()` or `filename()` on it.

**Comment by ctm** [comment]
I've commented out the call to `message()`. That may prevent the death that this issue is about, but without knowing why `message()` fails, it's impossible to know what other similar problems may lurk. So, I'm keeping this open for now.

**Comment by ctm** [comment]
I did a bunch of searching around yesterday, and did find a `wasm-bindgen` issue that was converted to a conversation that may be what I'm seeing. Now that I have a "minimal" (hah!) example of the problem, I've added [a comment](https://github.com/rustwasm/wasm-bindgen/discussions/3925#discussioncomment-10203316) to that conversation and am about to create a new issue in the web-sys repository.

**Comment by ctm** [comment]
Turns out, the web-sys repository _is_ the wasm-bindgen repository, so that's where my [new issue](https://github.com/rustwasm/wasm-bindgen/issues/4032) wound up.

**Comment by ctm** [comment]
Turns out, it was [a bug in `wasm-sockets`](https://github.com/scratchyone/wasm-sockets/issues/7).

I _still_ suspect this was a regression, at least on Chromium based browsers, but doing web searching turned up similar confusion from over a decade ago and my search-fu is just too weak for me to take the time to try to find something recent. OTOH, the error handling in mb2 is a bit complex and there's a chance that the places where I think I've seen the results of a successful `message()` call were just other errors that were handled properly that didn't involve that code path.

Either way, I've made an `error-7` branch of my fork of `wasm-sockets` and will use that for now.  I'm assuming the canonical branch will be fixed and when it is, I'll switch back, but I'm not leaving this issue open until then.
